service/test/_amqp_test.js

```diff
@@ -73,59 +67,87 @@ describe('Amqp', function() {
+      var amqp = new Amqp(fakeConfig, fakeAmqpBase);
+      amqp.on('disconnect', function() {
+        assert.fail();
+      });
+
+      fakeAmqpBase.emit('disconnect', new Error('fake'));
+    });
+
+    it('disconnects the transport if an AMQP error is emitted while connecting', function (testCallback) {
+      fakeAmqpBase.connect = sinon.stub();
+      var amqp = new Amqp(fakeConfig, fakeAmqpBase);
+      amqp.on('disconnect', function(err) {
+        assert.instanceOf(err, Error);
+        assert.isTrue(fakeAmqpBase.disconnect.calledOnce);
+        testCallback();
+      });
+
+      amqp.connect(function () {});
+      // blocked in connecting because the connect callback won't be called.
+      fakeAmqpBase.emit('disconnect', new Error('fake'));
+    });
+
+    it('disconnects the transport if an AMQP error is emitted while authenticating', function (testCallback) {
+      fakeAmqpBase.putToken = sinon.stub();
+      var amqp = new Amqp(fakeConfig, fakeAmqpBase);
+      amqp.on('disconnect', function(err) {
+        assert.instanceOf(err, Error);
+        assert.isTrue(fakeAmqpBase.disconnect.calledOnce);
+        testCallback();
+      });
+
+      amqp.connect(function () {});
+      // blocked in connecting because the putToken callback won't be called.
+      fakeAmqpBase.emit('disconnect', new Error('fake'));
     });
   });
 
   describe('#connect', function() {
     /*Tests_SRS_NODE_IOTHUB_SERVICE_AMQP_16_019: [The `connect` method shall call the `connect` method of the base AMQP transport and translate its result to the caller into a transport-agnostic object.]*/
     it('calls the base transport connect method', function(done) {
-      var amqp = new Amqp(fakeConfig, new FakeAmqpBase());
-      sinon.stub(amqp._amqp,'connect').callsArgWith(2, null);
+      var amqp = new Amqp(fakeConfig, fakeAmqpBase);
       amqp.connect(done);
     });
 
-    it('calls the base transport connect method which fails', function(done) {
+    /*Tests_SRS_NODE_IOTHUB_SERVICE_AMQP_16_017: [All asynchronous instance methods shall call the `done` callback with a single parameter that is derived from the standard Javascript `Error` object if the operation failed.]*/
+    it('calls its callback with an error if the base transport connect method fails', function(done) {
       var testError = new errors.NotConnectedError('fake error');
-      var amqp = new Amqp(fakeConfig, new FakeAmqpBase());
-      sinon.stub(amqp._amqp,'connect').callsArgWith(2, testError);
+      fakeAmqpBase.connect = sinon.stub().callsArgWith(2, testError);
+      var amqp = new Amqp(fakeConfig, fakeAmqpBase);
       amqp.connect(function (err) {
         assert.instanceOf(err, Error);
         done();
       });
     });
 
+    /*Tests_SRS_NODE_IOTHUB_SERVICE_AMQP_16_019: [The `connect` method shall call the `connect` method of the base AMQP transport and translate its result to the caller into a transport-agnostic object.]*/
```

Comment:
>/*Tests_SRS_NODE_IOTHUB_SERVICE_AMQP_16_019: [](start = 4, length = 44)

Not really sure what the test is.

Reply:
ugh. thanks.

---
In reply to: [147610681](https://github.com/Azure/azure-iot-sdk-node/pull/127#discussion_r147610681) [](ancestors = 147610681)